### PR TITLE
Searcher Controller with "this"

### PR DESCRIPTION
**Pull request by mwu-tow**
### Pull Request Description
Bringing searcher with node selected will use this node as a "this" argument for the function selected by the user.

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Rust](https://github.com/luna/enso/blob/main/docs/style-guide/rust.md).
- [x] All code has been tested where possible.
- [ ] All code has been profiled where possible.

### CODE REVIEW

File: src/rust/ide/src/model/suggestion_database.rs [hunk 1]
```diff
@@ -235,9 +237,9 @@ mod test {
             ..method_entry.clone()
         };
 
-        assert_eq!(atom_entry.code_to_insert()         , "Atom".to_string());
-        assert_eq!(method_entry.code_to_insert()       , "method".to_string());
-        assert_eq!(module_method_entry.code_to_insert(), "Main.moduleMethod".to_string());
+        assert_eq!(atom_entry.code_to_insert(None)         , "Atom".to_string());
+        assert_eq!(method_entry.code_to_insert(None)       , "method".to_string());
+        assert_eq!(module_method_entry.code_to_insert(None), "Main.moduleMethod".to_string());
```
Comment (farmaazon) [inline src/rust/ide/src/model/suggestion_database.rs:242]: Add cases for this being "non-none" ; .

File: src/rust/ide/src/model/suggestion_database.rs [hunk 2]
```diff
@@ -81,12 +81,14 @@ impl Entry {
...
 
     /// Returns the code which should be inserted to Searcher input when suggestion is picked.
-    pub fn code_to_insert(&self) -> String {
+    pub fn code_to_insert(&self, this_var:Option<&str>) -> String {
         let module = self.module.name();
         if self.self_type.as_ref().contains(&module) {
-            format!("{}.{}",module,self.name)
-        } else if self.self_type.as_ref().contains(&constants::keywords::HERE) {
+            format!("{}.{}",this_var.unwrap_or(module),self.name)
+        } else if self.self_type.as_ref().contains(&constants::keywords::HERE) { // TODO when this happens? why would the *type* be "here"?
```
Comment (farmaazon) [inline src/rust/ide/src/model/suggestion_database.rs:88]: My bad, the matter of `here` in suggestions should be discussed in an engine's issue somewhere, I'll do that and refer the issue here in my next PR.

File: src/rust/ide/src/controller/searcher.rs [hunk 3]
```diff
@@ -348,24 +419,53 @@ impl Searcher {
...
+                            error!(logger, "Failed to obtain type for this node.");
+                            None
+                        },
+                    }).boxed_local()
+                }
+                None => ready_boxed(None)
+            }
+        } else {
+            ready_boxed(None)
+        }
```
[the remark this answers is not in the document]
Comment (mwu-tow) [reply src/rust/ide/src/controller/searcher.rs:561]: Please see the new code. I went for... flattening!

File: src/rust/ide/src/controller/searcher.rs [hunk 4]
```diff
@@ -322,49 +343,49 @@ pub struct Searcher {
...
+    , selected_nodes : Vec<double_representation::node::Id>
+    ) -> FallibleResult<Self> {
+        let logger   = Logger::sub(parent,"Searcher Controller");
+        let database = project.suggestion_db();
+        let data     = if let Mode::EditNode{node_id} = mode {
+            Data::new_with_edited_node(&graph.graph(),&*database,node_id)?
+        } else {
+            default()
+        };
+        let this_arg = Rc::new(ThisNode::new(selected_nodes,&graph.graph()));
```
Comment (farmaazon) [inline src/rust/ide/src/controller/searcher.rs:377]: this_arg could be initialized directly in the structure.

File: src/rust/ide/src/controller/searcher.rs [hunk 5]
```diff
@@ -518,49 +562,37 @@ impl Searcher {
...
-                    }).boxed_local()
-                }
-                None => ready_boxed(None)
-            }
-        } else {
-            ready_boxed(None)
+    fn this_arg_type_for_next_completion(&self) -> impl Future<Output=Option<String>> {
+        let next_id = self.data.borrow().input.next_completion_id();
+        let logger  = self.logger.clone_ref();
+        let graph   = self.graph.clone();
```
Comment (farmaazon) [inline src/rust/ide/src/controller/searcher.rs:570]:
```suggestion
        let graph   = self.graph.clone_ref();
```

File: src/rust/lib/enso-prelude/src/option.rs [hunk 6]
```diff
@@ -6,6 +6,7 @@ pub trait OptionOps {
     fn map_ref      <U,F> (&self , f:F) -> Option<U> where F : FnOnce(&Self::Item) -> U;
     fn for_each     <U,F> (self  , f:F)              where F : FnOnce(Self::Item)  -> U;
     fn for_each_ref <U,F> (&self , f:F)              where F : FnOnce(&Self::Item) -> U;
+    fn map_none     <F>   (self  , f:F) -> Self      where F : FnOnce();
```
Comment (farmaazon) [inline src/rust/lib/enso-prelude/src/option.rs:9]: Why this is called "map"? It does not map anything. It's rather something like "if none"

Comment (mwu-tow) [reply src/rust/lib/enso-prelude/src/option.rs:9]: For symmetry with `Result` and `map_err`. Imagine that this function maps one `None` into another. As `None` is just a single value type, it is collapsed to void.

File: src/rust/ide/src/controller/searcher.rs (hunk 4 above)
Comment (mwu-tow) [reply src/rust/ide/src/controller/searcher.rs:377]: Tried but doesn't work well here, as in the structure the `graph` is moved. So we'd end up mixing both syntaxes for field initialization that doesn't look good.
Leaving as-is.

File: src/rust/ide/src/controller/searcher.rs [hunk 7]
```diff
@@ -590,6 +742,78 @@ mod test {
...
+                };
+
+                expect_call!(client.completion(
+                module      = data.graph.module.path.file_path().clone(),
+                position    = data.code_location,
+                self_type   = case.sets_this.as_some(mock_type.to_owned()),
+                return_type = None,
+                tag         = None
+            ) => Ok(completion_response));
+            });
```
Comment (wdanilo) [inline src/rust/ide/src/controller/searcher.rs:790]: very strange alignment.

Comment (mwu-tow) [reply src/rust/ide/src/controller/searcher.rs]: fixed

**Pull request merged**